daemon/api_prompting.go

```diff
@@ -97,6 +100,228 @@ func getUserID(r *http.Request) (uint32, Response) {
 	return uint32(userIDInt), nil
 }
 
+type invalidReason string
+
+const (
+	unsupportedValueReason invalidReason = "unsupported-value"
+	parseErrorReason       invalidReason = "parse-error"
+	validationErrorReason  invalidReason = "validation-error"
```

Comment:
is this distincition very relevant for the outside world?

Reply:
Yes, the idea Innes and I had was that different error reasons elicit different responses from the client.

- unsupported value errors have a known fixed set of valid values, so an error means that the client either has a bug or is incompatible with the current version of snapd, since it tried to send something which is not one of those values.
- parse errors mean that either the client or the user tried to send a value which is simply invalid in all cases (e.g. send a duration without a unit). So if the client constructed the response and got this error, the client has a bug, but if the user did (e.g. sent an invalid path pattern), then the client can perhaps show the user a link to documentation about how to write path patterns, and invite them to try again
- validation errors mean that the client/user submitted a well-formed response, but the current state of the system means that it wasn't accepted. The three errors of this type are:
  - prompt requested permissions [x, y, z], but user only responded with [x, z], so they should try again
  - prompt requested path `/foo/bar/baz` but the user replied with a path pattern which doesn't match it (e.g. `/foo/*/*/baz`), so they should try again
  - user tried to create a rule which conflicted with existing rules, so show the user the exact conflicts and ask them to try again

So in summary, we want the client to be able to differentiate whether:
- The client has a bug or is incompatible with current snapd
- The client/user submitted malformed data
- The client/user submitted well-formed data which was rejected.